>From 996ab259de5f6182e863893b27eea8af3e56c56c Mon Sep 17 00:00:00 2001
From: Daniel Hartwig <mandyke@gmail.com>
Date: Tue, 11 Sep 2012 12:40:46 +0800
Subject: [PATCH] Download changelogs from APT::Changelogs::Server

* src/generic/apt/pkg_changelog.cc:
  - download changelogs from APT::Changelogs::Server instead of
    a fixed location;
* src/apt_options.cc:
  - replace Aptitude::Changelog-URL-Template which was used only
    in ancient history with APT::Changelogs::Server
---
 NEWS                             |    3 +++
 src/apt_options.cc               |    8 ++++----
 src/generic/apt/pkg_changelog.cc |    9 ++++++++-
 3 files changed, 15 insertions(+), 5 deletions(-)

diff --git a/NEWS b/NEWS
index dffde9f..8eeecb1 100644
--- a/NEWS
+++ b/NEWS
@@ -17,6 +17,9 @@ Version 0.6.8.2                             "Say 'what' again"
                by default and works without deb-src entries.
                (Closes: #587775, #631464)
 
+  * [all]: Download changelogs from APT::Changelogs::Server instead
+           of a fixed location.
+
 - Documentation:
 
 - Translation updates:
diff --git a/src/apt_options.cc b/src/apt_options.cc
index bfeeda5..912aa09 100644
--- a/src/apt_options.cc
+++ b/src/apt_options.cc
@@ -291,12 +291,12 @@ option_item misc_options[]={
 		 "downloaded from any archive in sources.list."),
 	      PACKAGE "::AutoClean-After-Update", false),
   option_item(N_("URL to use to download changelogs"),
-	      N_("This option controls the template that's used to "
-		 "download changelogs from the Debian Web site.  You "
+	      N_("This option controls the server used to "
+		 "download changelogs.  You "
 		 "should only need to change this if the changelogs move "
 		 "to a different URL."),
-	      PACKAGE "::Changelog-URL-Template",
-	      "http://cgi.debian.org/cgi-bin/get-changelog?package=%s"),
+	      "APT::Changelogs::Server",
+	      "http://packages.debian.org/changelogs"),
   option_item(N_("Display a preview of what will be done before doing it"),
 	      N_("If this option is enabled, then when you ask "
 		 "aptitude to perform an install run, it will "
diff --git a/src/generic/apt/pkg_changelog.cc b/src/generic/apt/pkg_changelog.cc
index 9b1caf3..6e67251 100644
--- a/src/generic/apt/pkg_changelog.cc
+++ b/src/generic/apt/pkg_changelog.cc
@@ -20,6 +20,7 @@
 #include "pkg_changelog.h"
 
 #include "apt.h"
+#include "config_signal.h"
 #include "download_queue.h"
 
 #include <generic/util/job_queue_thread.h>
@@ -542,12 +543,18 @@ namespace aptitude
 	      else
 		realver = source_version;
 
-	      string uri = cw::util::ssprintf("http://packages.debian.org/changelogs/pool/%s/%s/%s/%s_%s/changelog",
+              // WATCH: apt/cmdline/apt-get.cc(DownloadChangelog)
+              string server = aptcfg->Find("APT::Changelogs::Server",
+                                           "http://packages.debian.org/changelogs");
+	      string path = cw::util::ssprintf("pool/%s/%s/%s/%s_%s",
 					      realsection.c_str(),
 					      prefix.c_str(),
 					      source_package.c_str(),
 					      source_package.c_str(),
 					      realver.c_str());
+              string uri = cw::util::ssprintf("%s/%s/changelog",
+                                              server.c_str(),
+                                              path.c_str());
 	      LOG_TRACE(logger,
 			"Adding " << uri
 			<< " as a URI for the changelog of " << source_package << " " << source_version);
-- 
1.7.9

